### PR TITLE
Adding new train_step logic to make things less confusing for users

### PR DESCRIPTION
**Pull request by Rocketknight1**
Change to `train_step` I discussed earlier with @gante, making the new TF approaches less confusing for users.

This PR is **extremely** experimental right now, and since it touches `train_step` it could break the entire TF side of the library, so please don't merge it until I'm done!

### CODE REVIEW

**Review by sgugger — APPROVED** [review]
Thanks for working on this!

File: tests/t5/test_modeling_tf_t5.py [hunk 1]
```diff
@@ -815,7 +815,7 @@ def __init__(self, name="accuracy", **kwargs):
                 super().__init__(_accuracy, name=name, **kwargs)
 
         model = self.model
-        model.compile("adam", metrics=FirstTokenAccuracy())
+        model.compile("adam", loss="sparse_categorical_crossentropy", metrics=FirstTokenAccuracy())
```
[the remark this answers is not in the document]
Comment (Rocketknight1) [reply tests/t5/test_modeling_tf_t5.py]: In this case I just dropped the test entirely! It overlaps very heavily with the new `test_keras_fit` common test, so I don't think we need to retain a T5-specific test when we have a common one.

File: tests/convnext/test_modeling_tf_convnext.py [hunk 2]
```diff
@@ -143,6 +143,10 @@ def setUp(self):
     def test_inputs_embeds(self):
         pass
 
+    @unittest.skip(reason="TF (<=2.8) does not support backprop for grouped convolutions on CPU.")
```
[the remark this answers is not in the document]
Comment (Rocketknight1) [reply tests/convnext/test_modeling_tf_convnext.py:146]: Done!

File: tests/test_modeling_tf_common.py [hunk 3]
```diff
@@ -1296,6 +1296,56 @@ def test_loss_computation(self):
 
                 self.assertEqual(loss.shape, [loss_size])
 
+    def test_keras_fit(self):
+        config, inputs_dict = self.model_tester.prepare_config_and_inputs_for_common()
+        for model_class in self.all_model_classes:
+            model = model_class(config)
+            if getattr(model, "hf_compute_loss", None):
```
Comment (gante) [inline tests/test_modeling_tf_common.py:1303]: Out of curiosity -- are there models w/o `hf_compute_loss`?

File: tests/test_modeling_tf_common.py [hunk 4]
```diff
@@ -1296,6 +1296,56 @@ def test_loss_computation(self):
...
+                possible_label_cols = {
+                    "labels",
+                    "label",
+                    "label_ids",
+                    "start_positions",
+                    "start_position",
+                    "end_positions",
+                    "end_position",
+                    "next_sentence_label",
+                }
```
Comment (gante) [inline tests/test_modeling_tf_common.py:1322]: This list is longer than the list in the train/test steps -- any reason behind the mismatch? 

(also, I'm surprised that these 8 names catch all the cases :o)

File: tests/test_modeling_tf_common.py [hunk 5]
```diff
@@ -1296,6 +1296,56 @@ def test_loss_computation(self):
...
+                history2 = model.fit(
+                    inputs_minus_labels,
+                    labels,
+                    validation_data=(inputs_minus_labels, labels),
+                    steps_per_epoch=1,
+                    validation_steps=1,
+                    shuffle=False,
+                )
+                val_loss2 = history2.history["val_loss"][0]
+                self.assertTrue(np.allclose(val_loss1, val_loss2, atol=1e-2, rtol=1e-3))
```
Comment (gante) [inline tests/test_modeling_tf_common.py:1347]: `1e-2` is a big difference, when it should be 0 🤔  What explains this difference?

**Review by gante — APPROVED** [review]
The next TF MLE will have an easier time with the loss handling, hopefully 🙏  LGTM!

Asked a few questions, to better understand what's going on in some parts of the code.

File: tests/test_modeling_tf_common.py (hunk 5 above)
Comment (gante) [reply tests/test_modeling_tf_common.py:1347]: hah, this test will likely break for the newly added `TFViTMAE`, since it adds noise in the forward pass 🙈

File: tests/test_modeling_tf_common.py (hunk 3 above)
Comment (Rocketknight1) [reply tests/test_modeling_tf_common.py:1303]: I don't believe so - this code might be vestigial. I'll check tomorrow!

File: tests/test_modeling_tf_common.py (hunk 5 above)
Comment (Rocketknight1) [reply tests/test_modeling_tf_common.py:1347]: Done!

**Comment by Rocketknight1** [comment]
Tests are passing and I think we're ready for final review!

I have a couple of things I'm still unsure about, though:

@gante: I removed the `expand_1d` call in `train_step`. Keras by default expands 1D input arrays to 2D, e.g. shape (8,) becomes (8, 1). This is not what our models expect at all and the test revealed some failures, that were fixed by removing it. Can you think of anything that change might break, though?

@sgugger / @LysandreJik: Is there any way to get a list of all possible 'label' column names for our models? This probably seems like a strange question, but the reason I need it is that if the user passes a single tensor as the labels to Keras, I need to make that tensor a key in the input dict, which means figuring out which argument name is the 'label'. Right now I just hardcoded a list of common ones, and then I inspect the function signature to see which of those is present, but it would be a lot better if I had some kind of definitive list.

**Comment by sgugger** [comment]
We don't have a definitive list and maintaining it is going to be difficult as well with new modalities arriving. Perhaps we should add an attribute `label_names` to all models to have this info stored somewhere? The `Trainer` would also benefit from it on the PyTorch side.

Wdyt @LysandreJik ?

**Comment by LysandreJik** [comment]
I wouldn't mind having the models know their label names, otherwise inspecting the signature and looking for any argument that contains `label` should give the correct information as well. If doing that during inference/training, it might slow down the forward pass so if you feel like adding a `label_names` is cleaner that's fine for me.

**Comment by sgugger** [comment]
The problem is that some models have labels without labels in their names (QA models, I'm looking at you!)

**Comment by Rocketknight1** [comment]
For TF models trained with Keras, the `call` (`forward`) method is usually only run once to build the graph, so expensive Python calls like `inspect` are totally okay!

That compilation step is really the source of a lot of different design decisions between the two frameworks - e.g. TF `einsum` takes a long time to figure out the optimal contraction during compilation and then saves it, so you can get very different runtimes as a result.

**Comment by LysandreJik** [comment]
> The problem is that some models have labels without labels in their names (QA models, I'm looking at you!)

For consistency's sake, shouldn't we rename them to have `label` in their name? (with appropriate deprecation warnings of course)
Maybe a non-issue if no users have ever been misled, but if it makes everything clearer *and* helps programmatic handling of labels, then it might be worth it

**Comment by sgugger** [comment]
I don't think making a change on the labels for the QA models is warranted as we can make an exception for those (double check the name "QuestionAnswering" and that it has `start_positions` and `end_positions` in its signature for instance). We can do the inspection once at init, so it's not a big deal in terms of performance.

**Comment by gante** [comment]
> @gante: I removed the expand_1d call in train_step. Keras by default expands 1D input arrays to 2D, e.g. shape (8,) becomes (8, 1). This is not what our models expect at all and the test revealed some failures, that were fixed by removing it. Can you think of anything that change might break, though?

I can't think of any problem. It is also being tested, so if there are problems, we should be able to catch them quickly.

**Comment by Rocketknight1** [comment]
For reference, here are all of the arguments for any `forward()` method in the codebase on a subclass of `PreTrainedModel` that contain "label" in their name:

`{'sentence_order_label', 'mc_labels', 'next_sentence_label', 'obj_labels', 'mask_labels', 'sentence_image_labels', 'class_labels', 'labels', 'entity_labels', 'matched_label'}`

@sgugger @LysandreJik None of them look especially wrong to me, so I'm happy to assume this list + {"start_positions", "end_positions"} covers all of the possible labels and just do an `inspect` call if you think that's a good solution!

**Comment by sgugger** [comment]
Added a utility in #16526

**Comment by Rocketknight1** [comment]
Quick update: This PR is mostly ready, but I'm going to wait for `find_labels` to be merged, then rebase and use it here.

**Comment by Rocketknight1** [comment]
Rebased with `find_labels` and looks good in testing, so I'm going to merge!
